### PR TITLE
IOS/ES: Merge the title import and export contexts

### PR DESCRIPTION
**Pull request by leoetlino**
This commit merges the import and export contexts into a single context because this is what IOS does, which means we can only reproduce its behaviour correctly if we use a single context for both operations. Overall, ES HLE should behave a lot more like ES during title imports now.

The other reason is that having two separate and very similar structs for no reason is not really a good idea.

Validity checks are also more explicit now (just need to check for `context.valid` instead of relying on a detail).
 
While working on this commit, I was notified that our handling of ImportTmd/ExportTitleInit is not correct. In particular, we always use the title key for both importing and exporting, which is wrong. To make this easier to fix in a follow-up PR, the context now also has a title key field, just like ES. This also lets us avoid computing it every single time in ImportContentDone.

### CODE REVIEW

File: Source/Core/Core/IOS/ES/ES.h [hunk 1]
```diff
@@ -55,35 +55,23 @@ class ES final : public Device
...
+    bool valid = false;
+    IOS::ES::TMDReader tmd;
+    std::array<u8, 16> key{};
+    struct ContentContext
     {
-      OpenedContent content;
+      bool valid = false;
+      u32 id = 0;
       std::array<u8, 16> iv{};
+      std::array<u8, 20> sha1{};
```
Comment (ligfx) [inline Source/Core/Core/IOS/ES/ES.h]: This isn't used anywhere, is this intentional?

File: Source/Core/Core/IOS/ES/ES.h [hunk 2]
```diff
@@ -55,35 +55,23 @@ class ES final : public Device
...
+      std::vector<u8> buffer;
+      u32 processed_size = 0;
```
Comment (ligfx) [inline Source/Core/Core/IOS/ES/ES.h]: This is incremented but never read, is this intentional?

File: Source/Core/Core/IOS/ES/TitleManagement.cpp [hunk 3]
```diff
@@ -45,6 +45,18 @@ static ReturnCode WriteTicket(const IOS::ES::TicketReader& ticket)
...
+void ES::TitleImportContext::DoState(PointerWrap& p)
+{
+  p.Do(valid);
+  p.Do(key);
+  tmd.DoState(p);
+  p.Do(content.valid);
+  p.Do(content.id);
+  p.Do(content.iv);
+  p.Do(content.sha1);
+  p.Do(content.buffer);
```
Comment (ligfx) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: Should `content.processed_size` be saved as well?

File: Source/Core/Core/IOS/ES/TitleManagement.cpp [hunk 4]
```diff
@@ -394,19 +402,25 @@ IPCCommandResult ES::ImportTitleDone(Context& context, const IOCtlVRequest& requ
 
 ReturnCode ES::ImportTitleCancel(Context& context)
 {
-  if (!context.title_import.tmd.IsValid())
+  // It is possible to get into a state where the TMD buffer exists and
```
Comment (ligfx) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: This comment looks like it's missing a part in the middle.

File: Source/Core/Core/IOS/ES/TitleManagement.cpp [hunk 5]
```diff
@@ -394,19 +402,25 @@ IPCCommandResult ES::ImportTitleDone(Context& context, const IOCtlVRequest& requ
...
+  // with the title import not being considered valid.
+  if (context.title_import.tmd.GetBytes().empty() || context.title_import.content.valid)
```
Comment (ligfx) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: Should this be `|| !context.title_import.content.valid`?

File: Source/Core/Core/IOS/ES/TitleManagement.cpp [hunk 6]
```diff
@@ -644,39 +654,40 @@ IPCCommandResult ES::ExportContentBegin(Context& context, const IOCtlVRequest& r
...
-
-  const u32 length =
-      std::min(static_cast<u32>(metadata.m_content.size - metadata.m_position), data_size);
-  std::vector<u8> buffer(length);
-
-  if (!content->m_Data->GetRange(metadata.m_position, length, buffer.data()))
-  {
-    ERROR_LOG(IOS_ES, "ExportContentData: ES_SHORT_READ");
-    return ES_SHORT_READ;
+    ret = [&] {
```
Comment (ligfx) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: I had a hard time following the flow of this function because of the lambda usage. IMO deduplicating the `CloseContent(content_fd, 0); context.title_import = {};` code isn't worth the hit to readability the lambda introduces.

File: Source/Core/Core/IOS/ES/TitleManagement.cpp [hunk 7]
```diff
@@ -644,39 +654,40 @@ IPCCommandResult ES::ExportContentBegin(Context& context, const IOCtlVRequest& r
 
 ReturnCode ES::ExportContentData(Context& context, u32 content_fd, u8* data, u32 data_size)
 {
-  const auto iterator = context.title_export.contents.find(content_fd);
-  if (!context.title_export.valid || iterator == context.title_export.contents.end() ||
-      iterator->second.content.m_position >= iterator->second.content.m_content.size)
+  ReturnCode ret = ES_EINVAL;
```
Comment (ligfx) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: I'm probably missing something, but this doesn't seem to error if `position >= size` anymore? Is that intentional?

File: Source/Core/Core/IOS/ES/TitleManagement.cpp [hunk 8]
```diff
@@ -696,20 +707,9 @@ IPCCommandResult ES::ExportContentData(Context& context, const IOCtlVRequest& re
 
 ReturnCode ES::ExportContentEnd(Context& context, u32 content_fd)
 {
-  const auto iterator = context.title_export.contents.find(content_fd);
-  if (!context.title_export.valid || iterator == context.title_export.contents.end() ||
-      iterator->second.content.m_position != iterator->second.content.m_content.size)
-  {
+  if (!context.title_import.valid || !context.title_import.content.valid)
```
Comment (ligfx) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: This no longer seems to error if the full content wasn't read (unless I missed something in `CloseContent`, which is highly likely). Is that intentional?

File: Source/Core/Core/IOS/ES/ES.h (hunk 1 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/ES.h]: Good catch -- I intended to change the import code to decrypt/hash each chunk of content in ImportContentData, instead of ImportContentEnd, to avoid keeping possibly large buffers in memory, but decided against putting it in this PR as it'd be an unrelated change.

File: Source/Core/Core/IOS/ES/ES.h (hunk 2 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/ES.h]: Looks like a leftover from debugging. I'll remove this.

File: Source/Core/Core/IOS/ES/TitleManagement.cpp (hunk 3 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: Since I've removed processed_size, we don't need to save it anymore.

File: Source/Core/Core/IOS/ES/TitleManagement.cpp (hunk 4 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: I've rewritten it to be shorter and clearer (hopefully).

File: Source/Core/Core/IOS/ES/TitleManagement.cpp (hunk 5 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: No, because we want to prevent the title import from being cancelled if there's still a content import in progress. Would you like me to mention this in the comment?

File: Source/Core/Core/IOS/ES/TitleManagement.cpp (hunk 7 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: Yes. That was needed in the old code because we used to read contents ourselves and check bounds manually, but this is now all handled by ReadContent, which will even make sure the proper error code is returned in several other error cases.

File: Source/Core/Core/IOS/ES/TitleManagement.cpp (hunk 6 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: Sure, I've removed the lambda.

File: Source/Core/Core/IOS/ES/TitleManagement.cpp (hunk 8 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: Yes. The old code added too many checks that weren't actually done by ES. This is the [ExportContentEnd](https://gist.github.com/leoetlino/ef700cce50930e65e40b7b748cb0ad50) function for the reference. Checks that are specific to the export code (context validity checks) are done here, whereas the cfd bound check is done in CloseContent.

You may have noticed that we reuse the content related functions, whereas IOS's export code just uses the fd directly -- this is partly because we don't emulate internal fd tables currently (and it should be equivalent anyway).

File: Source/Core/Core/IOS/ES/TitleManagement.cpp [hunk 9]
```diff
@@ -555,30 +567,32 @@ IPCCommandResult ES::DeleteContent(const IOCtlVRequest& request)
...
+  const auto ticket = DiscIO::FindSignedTicket(context.title_import.tmd.GetTitleId());
   if (!ticket.IsValid())
     return ES_NO_TICKET;
-  if (ticket.GetTitleId() != context.title_export.tmd.GetTitleId())
+  if (ticket.GetTitleId() != context.title_import.tmd.GetTitleId())
     return ES_EINVAL;
 
-  context.title_export.title_key = ticket.GetTitleKey(m_ios.GetIOSC());
+  // FIXME: this is wrong. The title key is *not* used here. Key #5 or a null key is.
+  context.title_import.key = ticket.GetTitleKey(m_ios.GetIOSC());
```
Comment (BhaaLseN) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: I assume this will get a follow-up PR or similar...what do you think of a stub implementation (that is currently wrong and contains the FIXME comment) that at least conveys the meaning of what should happen here?

File: Source/Core/Core/IOS/ES/ES.h [hunk 10]
```diff
@@ -92,8 +78,8 @@ class ES final : public Device
 
     u16 gid = 0;
     u32 uid = 0;
+    // The same context is used for both title imports and exports.
     TitleImportContext title_import;
```
Comment (BhaaLseN) [inline Source/Core/Core/IOS/ES/ES.h]: Not sure if leaving this as Title _Import_ Context makes too much sense when its also used for Export, but I'm not really convinced that simply calling it `TitleContext` (inside a class `Context`) is a sensible choice either. Can be ignored if we can't come up with a better name...

Comment (leoetlino) [inline Source/Core/Core/IOS/ES/ES.h]: Maybe title operation?

File: Source/Core/Core/IOS/ES/TitleManagement.cpp (hunk 9 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: Yes, it's fixed in #5739

File: Source/Core/Core/IOS/ES/TitleManagement.cpp (hunk 5 above)
Comment (ligfx) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: Nah, I think the logic's clear enough. I just wanted to check this was intentional because it replaced `!context.title_import.tmd.IsValid()`

File: Source/Core/Core/IOS/ES/TitleManagement.cpp (hunk 8 above)
Comment (ligfx) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: Okay, thanks for explaining it!

File: Source/Core/Core/IOS/ES/ES.h (hunk 10 above)
Comment (ligfx) [inline Source/Core/Core/IOS/ES/ES.h]: Yeah, this occurred to me too, but likewise I couldn't think of anything much better. Is something like `TitleImportExportContent title_import_export` too verbose?

File: Source/Core/Core/IOS/ES/TitleManagement.cpp [hunk 11]
```diff
@@ -644,39 +652,37 @@ IPCCommandResult ES::ExportContentBegin(Context& context, const IOCtlVRequest& r
...
 
-  if (!content->m_Data->GetRange(metadata.m_position, length, buffer.data()))
-  {
-    ERROR_LOG(IOS_ES, "ExportContentData: ES_SHORT_READ");
-    return ES_SHORT_READ;
+    const std::vector<u8> output =
+        Common::AES::Encrypt(context.title_import.key.data(),
+                             context.title_import.content.iv.data(), buffer.data(), buffer.size());
+    std::copy(output.cbegin(), output.cend(), data);
+    return ret;
```
Comment (ligfx) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: Hmm, I don't see where `ret` is being set anymore.

File: Source/Core/Core/IOS/ES/ES.h (hunk 10 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/ES.h]: Yes, I think this makes it a bit too long, sadly :(

File: Source/Core/Core/IOS/ES/TitleManagement.cpp (hunk 11 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: Oops, thanks for catching this! I've reorganised the function and removed `ret`.

Comment (ligfx) [inline Source/Core/Core/IOS/ES/TitleManagement.cpp]: LGTM 👍

File: Source/Core/Core/IOS/ES/ES.h (hunk 10 above)
Comment (leoetlino) [inline Source/Core/Core/IOS/ES/ES.h]: I changed `title_import` to `title_import_export` in https://github.com/dolphin-emu/dolphin/pull/5733/commits/1ba1b51606b2ed7f2c65da8733f59ab86b00561d, and it actually doesn't look too bad (and it makes it really clear that the same context is used for both operations). What do you think?
